### PR TITLE
cli: reorganise diagnostics subcommand

### LINKED ISSUE #5192: cli: Reorganize diagnostic subcommand(s)
There are a few purely diagnostic CLI subcommands that are generally useful, especially for Linkerd developers, but don't achieve any high-level user-facing tasks...

- `diagnostics`  Fetch metrics directly from the Linkerd control plane containers
- `endpoints`    Introspect Linkerd's service discovery state
- `metrics`      Fetch metrics directly from Linkerd proxies

I propose that we move all of these under `linkerd diagnostics` as follows:

- `linkerd diagnostics endpoints svc/foo [port]`

The `endpoints` command should become `linkerd diagnostics endpoints`. The authority-form of resolutions is somewhat of an implementation detail and, instead of a DNS-form name, the CLI should accept a service or pod reference and a port. Service lookups should be traffic-split aware, resolving via service profiles to obtain traffic splits before resolving individual leaf services.

- `linkerd diagnostics proxy-metrics -n NS po --selector ...` 

The `linkerd metrics` command should be renamed `linkerd diagnostics proxy-metrics`. It should accept the same CLI arguments as `kubectl get` (which it does not today); and it should dump the metrics from all selected resources.

- `linkerd diagnostics controller-metrics [-n NS] [workload]`

The `linkerd diagnostics` command, as it exists today, doesn't really fit into a world where controller is split across multiple namespaces. I.e., we already have a `linkerd-multicluster` namespace that isn't supported by the command. The command should be moved to `linkerd diagnostics controller-metrics`. By default it should dump the mertrics from the core control plane components, but it should also accept an optional workload reference (pod, deployment, label selector, etc) so that the command may be used with any linkerd controllers. I think we should assume/require that there is an `admin-http` container port for all control plane components, and this should be used to determine which port is read. (Alternately, there's no reason we need these ports to vary at all -- it seems easier/better to simply standardize on using the same port across all linkerd controllers, but this isn't really a requirement, i guess).

### PR DESCRIPTION
**Pull request by Pothulapati**
Fixes #5192, #5193

This PR moves `metrics`, `diagnostics`(which prints out metrics of
control-plane components), `endpoints` and `install-sp` into a new `diagnostics`
subcommand.

Signed-off-by: Tarun Pothulapati <tarunpothulapati@outlook.com>

### CODE REVIEW

File: cli/cmd/diagnostics.go [hunk 1]
```diff
@@ -14,29 +14,58 @@ const (
...
-func newDiagnosticsOptions() *diagnosticsOptions {
-	return &diagnosticsOptions{
+func newControlPlaneMetricsOptions() *ControlPlaneMetricsOptions {
+	return &ControlPlaneMetricsOptions{
 		wait: 30 * time.Second,
 	}
 }
 
-// newCmdDashboard creates a new cobra command `diagnostics` which contains commands to fetch control plane container's metrics
+// newCmdDiagnostics creates a new cobra command `diagnostics` which contains commands to diagnose Linkerd
```
Comment (kleimkuhler) [inline cli/cmd/diagnostics.go:33]:
```suggestion
// newCmdDiagnostics creates a new cobra command `diagnostics` which contains commands to fetch Linkerd diagnostics
```

File: cli/cmd/diagnostics.go [hunk 2]
```diff
@@ -14,29 +14,58 @@ const (
...
+}
+
+// newCmdControlPlaneMetrics creates a new cobra command `controlplane-metrics` which contains commands to fetch control plane container's metrics
+func newCmdControlPlaneMetrics() *cobra.Command {
+	options := newControlPlaneMetricsOptions()
 
 	cmd := &cobra.Command{
-		Use:   "diagnostics",
-		Short: "Fetch metrics directly from the Linkerd control plane containers",
+		Use:     "controlplane-metrics",
```
Comment (kleimkuhler) [inline cli/cmd/diagnostics.go:66]: We tend to use a space in "control plane" in comments, as well as the fact that we use the capitalization of "ControlPlane.." everywhere. Maybe this subcommand should be `control-plane-metrics`? It's a little more verbose which isn't great.

File: cli/cmd/diagnostics.go [hunk 3]
```diff
@@ -14,29 +14,58 @@ const (
...
+	}
+
+	diagnosticsCmd.AddCommand(newCmdControlPlaneMetrics())
+	diagnosticsCmd.AddCommand(newCmdEndpoints())
+	diagnosticsCmd.AddCommand(newCmdMetrics())
+	return diagnosticsCmd
+}
+
+// newCmdControlPlaneMetrics creates a new cobra command `controlplane-metrics` which contains commands to fetch control plane container's metrics
+func newCmdControlPlaneMetrics() *cobra.Command {
```
Comment (kleimkuhler) [inline cli/cmd/diagnostics.go:62]: What are your thoughts on maybe moving this to a `metrics.go`, or separating into a `control-plane-metrics.go` and `proxy-metrics.go`?

It feels a little odd to have the root `diagnotics` command here with one subcommand, but the other two subcommands defined in separate files.

File: cli/cmd/diagnostics.go (hunk 2 above)
Comment (Pothulapati) [reply cli/cmd/diagnostics.go:66]: From the issue, It is mentioned to be `controller-metrics` which seems concise. Will be doing that.

File: cli/cmd/diagnostics.go (hunk 3 above)
Comment (Pothulapati) [reply cli/cmd/diagnostics.go:62]: Yeah, That makes sense. Should we instead move all sub-commands of `diagnostics` into `diagnostics.go`,  like `multicluster.go` ? :thinking:

File: cli/cmd/diagnostics.go (hunk 2 above)
Comment (kleimkuhler) [reply cli/cmd/diagnostics.go]: Cool yep that looks good!

File: cli/cmd/diagnostics.go (hunk 3 above)
Comment (kleimkuhler) [reply cli/cmd/diagnostics.go]: Alex opened #5210 which basically is what we're discussing here. Maybe for now we can move all subcommands into `diagnostics.go` and address this as part of that?

Comment (adleong) [reply cli/cmd/diagnostics.go]: :+1: to @kleimkuhler's suggestion

File: cli/cmd/diagnostics.go [hunk 4]
```diff
@@ -1,78 +1,42 @@
...
+  # Install service profiles for the control-plane components.
+  linkerd diagnostics install-sp
+  `,
 	}
 
-	cmd.Flags().DurationVarP(&options.wait, "wait", "w", options.wait, "Time allowed to fetch diagnostics")
+	diagnosticsCmd.AddCommand(newCmdControllerMetrics())
+	diagnosticsCmd.AddCommand(newCmdEndpoints())
+	diagnosticsCmd.AddCommand(newCmdMetrics())
+	diagnosticsCmd.AddCommand(newCmdInstallSP())
```
Comment (Pothulapati) [inline cli/cmd/diagnostics.go:39]: Should we call the `install-sp` something else? As its a bit confusing as this installs service profiles for the control-plane components and not the service profiles crds itself

File: cli/cmd/diagnostics.go [hunk 5]
```diff
@@ -1,78 +1,42 @@
...
+	diagnosticsCmd := &cobra.Command{
+		Use:     "diagnostics [flags]",
+		Aliases: []string{"dg"},
+		Args:    cobra.NoArgs,
+		Short:   "Commands used to diagnose Linkerd components",
+		Long: `Commands used to diagnose Linkerd components.
+
+This command provides subcommands to diagnose the functionality of Linkerd.`,
+		Example: `  # Get control-plane component metrics
+  linkerd diagnostics cp-metrics
```
Comment (adleong) [inline cli/cmd/diagnostics.go:23]: cp-metrics doesn't seem like it's a valid abbreviation for this command.

**Review by adleong — APPROVED** [review]

File: cli/cmd/diagnostics.go (hunk 5 above)
Comment (kleimkuhler) [inline cli/cmd/diagnostics.go:23]: This can be updated now:

```suggestion
  linkerd diagnostics controller-metrics
```

File: cli/cmd/metrics.go [hunk 6]
```diff
@@ -62,13 +62,13 @@ func newCmdMetrics() *cobra.Command {
...
-  linkerd metrics po/pod-foo-bar
+  linkerd diagnostics metrics po/pod-foo-bar
 
   # Get metrics from the web deployment in the emojivoto namespace.
-  linkerd metrics -n emojivoto deploy/web
+  linkerd diagnostics metrics -n emojivoto deploy/web
 
   # Get metrics from the linkerd-controller pod in the linkerd namespace.
-  linkerd metrics -n linkerd $(
+  linkerd diagnostics metrics -n linkerd $(
```
Comment (kleimkuhler) [inline cli/cmd/metrics.go:71]: These should be `linkerd diagnostics proxy-metrics`.

**Review by kleimkuhler — APPROVED** [review]
This looks good. There are some doc updates you should change before merging, but changes make sense and work as expected.

**Pull request merged**